File: src/main/scala/spark/WordCount.scala
package spark

import org.apache.spark.{SparkConf, SparkContext}


object WordCount {

  def main(args: Array[String]): Unit = {
    /*val conf = new SparkConf().setMaster("spark://172.16.2.106:7077")
      .setAppName("WordCount")
      .setJars(Array("F:\\lx\\Code\\CodeStorage\\target\\CodeStorage-1.0-SNAPSHOT.jar"))
      .setIfMissing("spark.driver.host","172.16.2.56")*/
    //本地运行
    val conf = new SparkConf().setMaster("local[*]").setAppName("WordCount")
    //实例化spark上下文
    val sc = new SparkContext(conf)
    //读文件
    val textFile = sc.textFile("hdfs://172.16.2.106:8020/zx/spark/wordCount/words.txt")
    val result = textFile.flatMap(_.split(" ")).map((_,1)).reduceByKey(_ + _,1)
    result.collect().foreach(x => print(s"$x "))
    println()
    val sortResult = result.sortBy(x => x._2,ascending = false)
    sortResult.collect().foreach(x => print(s"$x "))
    sortResult.saveAsTextFile("F:\\lx\\Code\\CodeStorage\\target/result")
    sc.stop()
  }
}

File: bigdata/src/main/java/Spark/WordCount.java
package Spark;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.FlatMapFunction;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.api.java.function.PairFunction;
import scala.Tuple2;
import utils.FileUtil;

import java.util.Arrays;
import java.util.List;
import java.util.regex.Pattern;

/**
 * @Author: LX
 * @Date: 2019/3/27 19:31
 * @Version: 1.0
 */
public class WordCount {

    private static final Pattern SPACE = Pattern.compile(" ");

    public static void main(String[] args) {
        // 1：创建SparkConf对象,设置相关配置信息
        SparkConf sparkConf = new SparkConf().setAppName("JavaWordCount");
        sparkConf.setMaster("local");

        // 2：创建JavaSparkContext对象，SparkContext是Spark的所有功能的入口
        JavaSparkContext ctx = new JavaSparkContext(sparkConf);

        // 3：创建一个初始的RDD
        // SparkContext中，用于根据文件类型的输入源创建RDD的方法，叫做textFile()方法
        JavaRDD<String> lines = ctx.textFile("D:\\Software\\spark-2.4.0-bin-hadoop2.7\\README.md", 1);
        // 4：对初始的RDD进行transformation操作，也就是一些计算操作
        StringBuilder sb = new StringBuilder();

        JavaRDD<String> words = lines.flatMap(
                (FlatMapFunction<String, String>) s -> Arrays.asList(SPACE.split(s)).iterator());

        JavaPairRDD<String, Integer> ones = words.mapToPair(
                (PairFunction<String, String, Integer>) s -> new Tuple2<>(s, 1));
        JavaPairRDD<String, Integer> counts = ones.reduceByKey(
                (Function2<Integer, Integer, Integer>) (i1, i2) -> i1 + i2);

        List<Tuple2<String, Integer>> output = counts.collect();

        // 5:将结果输出文件或者打印
        FileUtil util = new FileUtil();

        for (Tuple2<?, ?> tuple : output) {
//            System.out.println(tuple._1() + ": " + tuple._2());
            sb.append(tuple._1).append(':').append(tuple._2).append('\n');
        }
        util.writeString(sb.toString(),"d:/tmp/output/wordCount.txt");
        ctx.close();
    }
}
